include/exadg/incompressible_navier_stokes/postprocessor/postprocessor.cpp

```diff
@@ -195,116 +237,142 @@ PostProcessor<dim, Number>::do_postprocessing(VectorType const &     velocity,
+      };
+      streamfunction.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
+        navier_stokes_operator->compute_streamfunction(dst, src);
+      };
+
+      streamfunction.reinit();
+    }
   }
 
   // divergence
   if(pp_data.output_data.write_divergence)
   {
-    divergence.type        = SolutionFieldType::vector;
-    divergence.name        = "div_u";
-    divergence.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    navier_stokes_operator->initialize_vector_velocity_scalar(divergence.get_vector_reference());
+    divergence.type              = SolutionFieldType::vector;
+    divergence.name              = "div_u";
+    divergence.dof_handler       = &navier_stokes_operator->get_dof_handler_u_scalar();
+    divergence.initialize_vector = [&](std::shared_ptr<VectorType> & dst) {
+      dst = std::make_shared<VectorType>();
+      navier_stokes_operator->initialize_vector_velocity_scalar(*dst);
+    };
     divergence.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
       navier_stokes_operator->compute_divergence(dst, src);
     };
 
-    this->additional_fields_vtu.push_back(&divergence);
+    divergence.reinit();
   }
 
   // velocity magnitude
   if(pp_data.output_data.write_velocity_magnitude)
   {
-    velocity_magnitude.type        = SolutionFieldType::scalar;
-    velocity_magnitude.name        = "velocity_magnitude";
-    velocity_magnitude.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    navier_stokes_operator->initialize_vector_velocity_scalar(
-      velocity_magnitude.get_vector_reference());
+    velocity_magnitude.type              = SolutionFieldType::scalar;
+    velocity_magnitude.name              = "velocity_magnitude";
+    velocity_magnitude.dof_handler       = &navier_stokes_operator->get_dof_handler_u_scalar();
+    velocity_magnitude.initialize_vector = [&](std::shared_ptr<VectorType> & dst) {
+      dst = std::make_shared<VectorType>();
+      navier_stokes_operator->initialize_vector_velocity_scalar(*dst);
+    };
     velocity_magnitude.recompute_solution_field = [&](VectorType & dst, VectorType const & src) {
       navier_stokes_operator->compute_velocity_magnitude(dst, src);
     };
 
-    this->additional_fields_vtu.push_back(&velocity_magnitude);
-  }
-
-  // vorticity magnitude
-  if(pp_data.output_data.write_vorticity_magnitude)
-  {
-    vorticity_magnitude.type        = SolutionFieldType::scalar;
-    vorticity_magnitude.name        = "vorticity_magnitude";
-    vorticity_magnitude.dof_handler = &navier_stokes_operator->get_dof_handler_u_scalar();
-    navier_stokes_operator->initialize_vector_velocity_scalar(
-      vorticity_magnitude.get_vector_reference());
-    vorticity_magnitude.recompute_solution_field = [&](VectorType & dst, VectorType const &) {
-      navier_stokes_operator->compute_vorticity_magnitude(dst, evaluate_get(vorticity));
```

Comment:
This should be avoided. In my opinion it is hard to see that the vector passed to `evaluate()`/`evaluate_get()` is by-passed in case of `vorticity_magnitude`, given that the `reinit_additional_fields()` function says `vorticity_magnitude.reinit(velocity);` (velocity, not vorticity!).

Reply:
Before, `vorticity_magnitude` received `velocity` in the reinit function (`vorticity_magnitude.reinit(velocity)`), but the above lambda uses `vorticity` as source vector. This is why I proposed to remove the vector from the reinit-functions and explicitly pass the src-vector to all `evaluate()/evaluate_get()` functions. I do not see the disadvantage that the `evaluate()/evaluate_get()` functions have a parameter. This way, the "flow of information" and the dependency between derived fields becomes clear. I can certainly make a mistake by passing the wrong vector. I argue, however, that it is easier to find this mistake that in the current situation where the lambda accesses some vectors where I do not fully understand where they come from or when they are updated.